### PR TITLE
Add example for amp-font

### LINKED ISSUE #243: Create an example showing how amp-font works

### PR DESCRIPTION
**Pull request by kul3r4**
Fixes #243

### CODE REVIEW

File: src/20_Components/amp-font.html [hunk 1]
```diff
@@ -0,0 +1,65 @@
+<!-- 
+  #### Introduction
+
+  `amp-font` allows to control timeouts on font loading on AMP HTML pages.
+
+  AMP HTML files don't support the normal HTML `img` tag. With `amp-img` AMP
+  provides a powerful replacement.
```
Comment (sebastianbenz) [inline src/20_Components/amp-font.html]: Is this intended?

File: src/20_Components/amp-font.html [hunk 2]
```diff
@@ -0,0 +1,65 @@
+<!-- 
+  #### Introduction
+
+  `amp-font` allows to control timeouts on font loading on AMP HTML pages.
```
Comment (sebastianbenz) [inline src/20_Components/amp-font.html]: We should explain what amp-font is good for, maybe: "Learn how to use the `amp-font` component to trigger and monitor the loading of custom fonts on AMP pages."

File: src/20_Components/amp-font.html [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  -->
+  <p class="comic-amp">
+    The following text will be displayed in red if the chosen font is not supported.
+  </p>
+  <p class="comic-amp">
+    "Neque porro quisquam est qui dolorem ipsum quia dolor sit amet, consectetur, adipisci velit..."
+    "There is no one who loves pain itself, who seeks after it and wants to have it, simply because it is pain..."
+  </p>
+
+  <amp-font
```
Comment (sebastianbenz) [inline src/20_Components/amp-font.html]: Can you explain how to embed amp-font? What does the attributes mean? Does it need to be in the body?

File: src/20_Components/amp-font.html [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    }
+  </style>
+</head>
+<body class="comic-amp-font-loading comic-amp-bold-font-loading">
+  <!-- #### Basic Usage -->
+  <!--
+    A simple text showing how to use `amp-font`.
+  -->
+  <p class="comic-amp">
+    The following text will be displayed in red if the chosen font is not supported.
```
Comment (sebastianbenz) [inline src/20_Components/amp-font.html]: You could get away with using only a single paragraph if you rephrase to: "This text is displayed in red if the chosen font is not supported."

File: src/20_Components/amp-font.html [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  <script async custom-element="amp-font" src="https://cdn.ampproject.org/v0/amp-font-0.1.js"></script>
+  <!-- 
+    Define some CSS classes so that `amp-font` can add or remove those based on whether a font was loaded or is in error-state. Find out about all the supported css classes [here](https://github.com/ampproject/amphtml/blob/b1607dff9446d332afcdead64b63e8d25e74cff7/extensions/amp-font/amp-font.md)
+  -->
+  <style amp-custom>
+    @font-face {
+      font-family: 'Comic AMP';
+      font-style: normal;
+      font-weight: 400;
+      src: url(fonts/ComicAMP.ttf) format('truetype');
```
Comment (sebastianbenz) [inline src/20_Components/amp-font.html]: ComicAMP.ttf isn't suppose to exist - correct? It's a bit confusing - can you make it more explicit? e.g. FontDoesNotExist.ttf

File: src/20_Components/amp-font.html [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    }
+    .comic-amp-font-loading .comic-amp {
+      color: #0ff;
+    }  
+    .comic-amp-font-missing .comic-amp {
+      color: red;
+    }
+  </style>
+</head>
+<body class="comic-amp-font-loading comic-amp-bold-font-loading">
```
Comment (sebastianbenz) [inline src/20_Components/amp-font.html]: Are we using comic-amp-bold-font-loading ?

**Comment by sebastianbenz** [comment]
Added  a few comments. Can we also demonstrate the good case?

File: src/20_Components/amp-font.html (hunk 1 above)
Comment (kul3r4) [inline src/20_Components/amp-font.html]: Nope, copy and paste trap!

File: src/20_Components/amp-font.html (hunk 2 above)
Comment (kul3r4) [inline src/20_Components/amp-font.html]: Ok

File: src/20_Components/amp-font.html (hunk 3 above)
Comment (kul3r4) [inline src/20_Components/amp-font.html]: Added some more explanation. I would avoid to copy the description of all the attributes and refer to the official docs, it's more maintanable.

File: src/20_Components/amp-font.html (hunk 4 above)
Comment (kul3r4) [inline src/20_Components/amp-font.html]: Ok

File: src/20_Components/amp-font.html (hunk 5 above)
Comment (kul3r4) [inline src/20_Components/amp-font.html]: Ok

**Comment by kul3r4** [comment]
@sebastianbenz Thanks for your review and comments.  PTAL

File: src/20_Components/amp-font.html [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  </p>
+    <!--
+    Embed amp-font inside the body or the head of the document.
+    Use the attribute `timeout` to specify the time in milliseconds to wait until the font becomes available. This attribute is optional and it's default value is 3000.
+    Use `on-error-remove-class`, `on-error-add-class`, `on-load-remove-class`, `on-load-add-class` to specify the CSS classes to be used if the font loads correctly or not. For a detailed description,, please refer to the [documentation](https://www.ampproject.org/docs/reference/extended/amp-font.html)
+    -->
+  <amp-font
+    layout="nodisplay"
+    font-family="Font Does Not exist"
+    timeout="2000"
```
Comment (cramforce) [inline src/20_Components/amp-font.html]: Would be good to have an example with `timeout=0`. This is the coolest mode: "Load only if already in cache"

File: src/20_Components/amp-font.html [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  <script async src="https://cdn.ampproject.org/v0.js"></script>
+  <script async custom-element="amp-font" src="https://cdn.ampproject.org/v0/amp-font-0.1.js"></script>
+  <!-- 
+    Define some CSS classes so that `amp-font` can add or remove those based on whether a font was loaded or is in error-state. Find out about all the supported css classes [here](https://github.com/ampproject/amphtml/blob/b1607dff9446d332afcdead64b63e8d25e74cff7/extensions/amp-font/amp-font.md)
```
Comment (sebastianbenz) [inline src/20_Components/amp-font.html]: "some CSS classes" is very unspecific. Maybe: "Define custom CSS rules to style your document  depending on whether a font was loaded successfully or not."

File: src/20_Components/amp-font.html [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+      font-style: normal;
+      font-weight: 400;
+      src: url(fonts/FontDoesNotExist.ttf) format('truetype');
+    }
+    @font-face {
+      font-family: 'Tangerine';
+      font-style: normal;
+      font-weight: 400;
+      src: local('Tangerine'), url(http://fonts.gstatic.com/s/tangerine/v7/HGfsyCL5WASpHOFnouG-RJBw1xU1rKptJj_0jans920.woff2) format('woff2');
+      unicode-range: U+0000-00FF, U+0131, U+0152-0153, U+02C6, U+02DA, U+02DC, U+2000-206F, U+2074, U+20AC, U+2212, U+2215, U+E0FF, U+EFFD, U+F000;
```
Comment (sebastianbenz) [inline src/20_Components/amp-font.html]: Remove `unicode-range` as it is not required.

File: src/20_Components/amp-font.html [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    .font-does-not-exist-loading {
+      color: #00f;
+    }
+    .font-does-not-exist-loading .font-does-not-exist {
+      color: #0ff;
+    }  
+    .font-does-not-exist-missing .font-does-not-exist {
+      color: red;
+    }
+    .slabo-loaded .slabo-exist{
```
Comment (sebastianbenz) [inline src/20_Components/amp-font.html]: what does slabo mean? Maybe use a better name.

File: src/20_Components/amp-font.html [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    }
+    @font-face {
+      font-family: 'Tangerine';
+      font-style: normal;
+      font-weight: 400;
+      src: local('Tangerine'), url(http://fonts.gstatic.com/s/tangerine/v7/HGfsyCL5WASpHOFnouG-RJBw1xU1rKptJj_0jans920.woff2) format('woff2');
+      unicode-range: U+0000-00FF, U+0131, U+0152-0153, U+02C6, U+02DA, U+02DC, U+2000-206F, U+2074, U+20AC, U+2212, U+2215, U+E0FF, U+EFFD, U+F000;
+    }
+
+    .font-does-not-exist-loaded .font-does-not-exist{
```
Comment (sebastianbenz) [inline src/20_Components/amp-font.html]: formatting nit: `.font-does-not-exist{` => `.font-does-not-exist {`

**Comment by sebastianbenz** [comment]
* You are demonstrating different amp-font scenarios. Can you structure the sample accordingly? 
* The sample text doesn't use the Tangerine font even if it is loaded correctly
* Is it necessary to duplicate the error/loading css rules? Can we just define them once and reuse them for both fonts?

File: src/20_Components/amp-font.html (hunk 7 above)
Comment (sebastianbenz) [inline src/20_Components/amp-font.html]: +1

**Comment by kul3r4** [comment]
Addressed the comments. @sebastianbenz PTAL

File: src/20_Components/amp-font.html (hunk 7 above)
Comment (kul3r4) [inline src/20_Components/amp-font.html]: Thanks!

File: src/20_Components/amp-font.html [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+      color: #0ff;
+    }  
+    .font-does-not-exist-missing .font-does-not-exist, .tangerine-missing .tangerine, .another-font-does-not-exist-missing .another-font-does-not-exist  {
+      color: red;
+    }
+
+  </style>
+</head>
+<body>
+  <!-- #### Sample for a not existing font -->
```
Comment (sebastianbenz) [inline src/20_Components/amp-font.html]: s/not existing/non-existing/

File: src/20_Components/amp-font.html [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+    timeout="2000"
+    on-error-remove-class="tangerine-loading"
+    on-error-add-class="tangerine-missing"
+    on-load-remove-class="tangerine-loading"
+    on-load-add-class="tangerine-loaded">
+  </amp-font>
+  <p class="tangerine">
+    This text is displayed in red if the chosen font is not supported.
+  </p>
+  <!-- #### Sample for using only cache font -->
```
Comment (sebastianbenz) [inline src/20_Components/amp-font.html]: s/Sample for using only cache font/Using only cached fonts/

File: src/20_Components/amp-font.html [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+      font-family: 'Font Does Not exist', serif, sans-serif;
+    }
+    .tangerine-loaded .tangerine {
+      font-family: 'Tangerine', serif, sans-serif;
+    }
+    .another-font-does-not-exist-loaded .another-font-does-not-exist {
+      font-family: 'Another Font That Does Not exist', serif, sans-serif;
+    }
+    .font-does-not-exist-loading, .tangerine-loading, .another-font-does-not-exist-loading {
+      color: #00f;
```
Comment (sebastianbenz) [inline src/20_Components/amp-font.html]: use `color: blue` if you're using red below

File: src/20_Components/amp-font.html [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+    }
+    .font-does-not-exist-loading .font-does-not-exist, .tangerine-loading .tangerine, .another-font-does-not-exist-loading .another-font-does-not-exist {
+      color: #0ff;
```
Comment (sebastianbenz) [inline src/20_Components/amp-font.html]: use `color: aqua` if you're using red below

File: src/20_Components/amp-font.html [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+    .tangerine-loaded .tangerine {
+      font-family: 'Tangerine', serif, sans-serif;
+    }
+    .another-font-does-not-exist-loaded .another-font-does-not-exist {
+      font-family: 'Another Font That Does Not exist', serif, sans-serif;
+    }
+    .font-does-not-exist-loading, .tangerine-loading, .another-font-does-not-exist-loading {
+      color: #00f;
+    }
+    .font-does-not-exist-loading .font-does-not-exist, .tangerine-loading .tangerine, .another-font-does-not-exist-loading .another-font-does-not-exist {
```
Comment (sebastianbenz) [inline src/20_Components/amp-font.html]: formatting nit: I think formatting it like this makes it easier to read:

```
font-does-not-exist-loading .font-does-not-exist, 
.tangerine-loading .tangerine,
 .another-font-does-not-exist-loading .another-font-does-not-exist {
     color: #0ff;
}
```

Same applies to the rule below.

**Comment by sebastianbenz** [comment]
suggestion: rename .font-does-not-exist -> .unavailable-font

File: src/20_Components/amp-font.html [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+    layout="nodisplay"
+    font-family="Another Font That Does Not exist"
+    timeout="0"
+    on-error-remove-class="another-font-does-not-exist-loading"
+    on-error-add-class="another-font-does-not-exist-missing"
+    on-load-remove-class="another-font-does-not-exist-loading"
+    on-load-add-class="another-font-does-not-exist-loaded">
+  </amp-font>
+  <p class="another-font-does-not-exist">
+    This text is displayed in red if the chosen font is not supported.
```
Comment (sebastianbenz) [inline src/20_Components/amp-font.html]: s/supported/cached/

File: src/20_Components/amp-font.html [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+    on-error-add-class="tangerine-missing"
+    on-load-remove-class="tangerine-loading"
+    on-load-add-class="tangerine-loaded">
+  </amp-font>
+  <p class="tangerine">
+    This text is displayed in red if the chosen font is not supported.
+  </p>
+  <!-- #### Sample for using only cache font -->
+  <!--
+  Set timeout parameter to 0 to load the font only if already in cache.
```
Comment (sebastianbenz) [inline src/20_Components/amp-font.html]: s/timeout parameter to 0/`timeout="0"`/

File: src/20_Components/amp-font.html (hunk 15 above)
Comment (kul3r4) [inline src/20_Components/amp-font.html]: To which class exactly?

Comment (sebastianbenz) [inline src/20_Components/amp-font.html]: s/color: #0ff/color: aqua/

**Comment by kul3r4** [comment]
@sebastianbenz PTAL

**Comment by sebastianbenz** [comment]
LGTM - so happy that we finally have a sample for amp-font :-)

**Pull request merged**